### PR TITLE
Fix 5676 specify branch layout

### LINKED ISSUE #5676: Incorrect layout of the 'Specify branch' dialog
**Do you want to request a *feature* or report a *bug*?**
A bug.

**What is the current behavior?**
The list of branches in the _Specify branch_ dialog scrolls horizontally instead of vertically and cannot be made larger.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
1. Add a user script which uses the `{sBranch}` argument
2. Create several branches with long-ish names at a single revision
3. Run the script on that revision

**What is the expected behavior?**
The _Specify branch_ dialog allows an easy selection of a branch. Specifically:
1. The branch list should scroll vertically
2. When the dialog is resized, the list should be resized, too

The images below demonstrate the issues.

obrazek

obrazek

**Environment you encounter the issue:**
- Git Extensions 3.00.a1
- ca83a30227057fe263e9fe53656c165ca292247f 
- Git 2.19.1.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3132.0

**Did this work in previous version of GitExtensions (which)?**
I don't think so.

### PR DESCRIPTION
**Pull request by crypto-rsa**
Fixes #5676.

Changes proposed in this pull request:
- layout logic extracted from `FormQuickGitRefSelector` into new base class `FormQuickItemSelector`
- added `Select` action to `FormQuickGitRefSelector`
- added new class `FormQuickStringSelector` for selecting from a collection of arbitrary strings
- `ScriptRunner` now shows the quick selector form instead of old `FormRunScriptSpecify` which has been removed
 
Screenshots before and after (if PR changes UI):
Before:
before

After:
after

What did I do to test the code and ensure quality:
- manually tested selecting branches and tags

Has been tested on (remove any that don't apply):
- Git 2.19.1.windows.1
- Microsoft Windows NT 10.0.17134.0

### CODE REVIEW

**Comment by RussKie** [comment]
Looking at the screenshot, I propose few tweaks:
* remove "_Specify 'Selected...'_" label, the form title is sufficient
* rename the form title to "_Select branch_"

Also does it need to be a listview control? Is use expected to select more than one branch?
If not, perhaps a dropdownlist is a better control for the job.

**Comment by crypto-rsa** [comment]
The dialog is used for selecting a branch, a tag or a remote (see `ScriptRunner.AskToSpecify`). The label text is determined by the caller.
As for the listview vs. dropdown - it only returns the first selected item, so the latter control is probably a better choice.

File: GitUI/HelperDialogs/FormRunScriptSpecify.Designer.cs [hunk 1]
```diff
@@ -39,24 +39,34 @@ private void InitializeComponent()
...
             this.branchesListView.Name = "branchesListView";
             this.branchesListView.Size = new System.Drawing.Size(290, 52);
             this.branchesListView.TabIndex = 1;
             this.branchesListView.UseCompatibleStateImageBehavior = false;
-            this.branchesListView.View = System.Windows.Forms.View.List;
+            this.branchesListView.View = System.Windows.Forms.View.Details;
+            // 
+            // dummyColumnHeader
+            // 
+            this.dummyColumnHeader.Text = "<dummy column>";
```
Comment (gerhardol) [inline GitUI/HelperDialogs/FormRunScriptSpecify.Designer.cs]: The translation app may pick up this.

Comment (crypto-rsa) [reply GitUI/HelperDialogs/FormRunScriptSpecify.Designer.cs]: I guess I should leave the column text empty. It is not visible anyway and the text is not used for sizing, either.

**Comment by crypto-rsa** [comment]
> Also does it need to be a listview control? Is use expected to select more than one branch?
> If not, perhaps a dropdownlist is a better control for the job.

I had a second thought about this and now I am not so sure about the dropdown. The thing is that there will very likely be just a small number of items (~5) in the list. It's either the remotes, or branch heads/tags _at a single revision_. In such case, isn't it better to have all the items shown right when the dialog opens instead of having to click the dropdown?

**Comment by RussKie** [comment]
One of the key drivers for me is consistent UX, so if the proposal aligns with everything else (or majority) we have designed - I'm content.

I wasn't aware of the form until this ticket, so I don't have a strong opinion about it. 
The proposal is certainly a significant improvement, but feels somewhat different to other forms we have. Most form dealing with branches and tags have dropdowns (e.g. delete branch, pull, push) or single line textboxes (e.g. create branch/tag, rename branch tag).
There is, of course, a form for selecting branch/tag to rename/delete when triggered from the keyboard, which is a listview. But it is more an exception than rule.

I went to check whether there was a UX opinion on _listview vs combobox_ and... I couldn't find one. There was [one article](https://blog.prototypr.io/7-rules-of-using-radio-buttons-vs-drop-down-menus-fddf50d312d1) that come close, and it seemed interesting.
I'm interested in your opinion, hence I asked the question 😄  

I know, this form is not _that_ important, but this maybe an opportunity to discuss and test different hypothesis and approaches. Who knows, we may apply learnings to the high visibility components/forms later.

**Comment by RussKie** [comment]
Feel free to to reuse it for the use case.
IIRC it is shown without a control box though, this may need to be made
configurable.

**Comment by crypto-rsa** [comment]
> IIRC it is shown without a control box though, this may need to be made configurable.

Do you think so? I actually don't see a reason why it should look or behave differently than in the Delete/Rename scenario (except for the button text).

**Comment by RussKie** [comment]
I'll let you run with it, see what you can come up with

**Comment by crypto-rsa** [comment]
Hm, there is one thing in which the current dialog differs from the Delete/Rename one - the label above the list which says what is being selected. Here are all the strings being used:
Current Revision Tag
Current Revision Branch
Current Revision Local Branch
Current Revision Remote Branch
Selected Revision Tag
Selected Revision Branch
Selected Revision Local Branch
Selected Revision Remote Branch
Selected Revision Remote

Honestly, I don't think it is useful at all. It is clear from the list whether a branch or a tag is being selected and whether a branch is local or remote. So perhaps we could drop it altogether.

**Comment by crypto-rsa** [comment]
One more thing: I guess I am becoming an expert in making pull requests which soon become obsolete 😢 Should I close this one and try the different approach in a separate PR?

**Comment by RussKie** [comment]
From the original screenshot I wouldn't think that label was particularly
useful, hence I proposed to remove it.
If you don't think it is really needed - remove it.

> I guess I am becoming an expert in making pull requests which soon become
obsolete

Unfortunately some times we have to pay this tax for exploratory work.

>  Should I close this one and try the different approach in a separate PR?

If a new PR would be addressing the same issue, I see no reason to open a
new one. Just reset this one and force push. That's how I'd do it.

**Comment by crypto-rsa** [comment]
Force pushed v2 of the PR. The revision is now selected using the `FormQuickGitRefSelector` so the UI is the same as when renaming or deleting a branch/tag. A new `FormQuickStringSelector` uses the same UI for letting the user select a string (here used for selecting a remote).

**Comment by RussKie** [comment]
Awesome, I'll try to review over the weekend.

**Comment by crypto-rsa** [comment]
Crossing my fingers for _this_ weekend :wink:

**Comment by RussKie** [comment]
@crypto-rsa my apologies, RL commitments are taking precedence over PR reviews atm.
There is a mounting pressure to do a release, so this work will go in to a subsequent maintenance release. Sorry.

**Comment by crypto-rsa** [comment]
No problem at all, I was just kidding. I understand this is not a release blocker 😄

**Comment by RussKie** [comment]
@crypto-rsa It appears my latest refactor of the `ScriptRunner` caused a MC on your PR. 
Could you please resolve it and we'll get it merged

**Comment by crypto-rsa** [comment]
> Could you please resolve it and we'll get it merged

Done. I had to add parameter `revisionGrid` to `ParseScriptArguments`. I wasn't happy to do it given the number of existing parameters in the method (eighteen!) but I think that's a separate issue.

**Comment by vbjay** [comment]
Instead of parameters, how about a struct/class so we can just add to the
struct/class instead of adding parameters?  Same principle as eventargs.

**Comment by RussKie** [comment]
Thank you.
The signature is horrible, agree. We'll need to look at it separately, my original refactor was aimed at making the code testable.
Thinking out loud, I'd probably change the code that each individual parameter is handled in a separate method, have a number of script argument substitution classes (one for each arg) and a factory class...

**Pull request merged**